aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -148,6 +169,10 @@ SparseTensor new_with_tensor_and_size_sparse(const LongTensor& indices, const Te
   }
 
   const SparseType& dtype = values.type().toSparse();
+
+#ifdef USE_TH_SIZE_ZERO_DIM
+  AT_CHECK(!(sizes.size() == 0 && values.numel() == 0), "cannot construct sparse tensor with empty size and no values; use the nullary constructor instead");
```

Comment:
This doesn't seem right (there's another place above it too that's similar), because you are throwing away the information about the sparse/dense dimensions (in general I'd be wary of anything that branches on numel() == 0), that is probably some special case handling that you don't need anymore.

For example, I compiled your branch with USE_TH_SIZE_ZERO_DIM and I get the following:
```
torch.sparse_coo_tensor(torch.empty((1,0), dtype=torch.int64), torch.empty(0,2,4,0), size=(0,2,4,0))
>>> torch.sparse.FloatTensor of size (0,2,4,0) with indices:
tensor([], size=(4, 0), dtype=torch.int64)
and values:
tensor([])
```
 which seems clearly wrong.

Reply:
I did some clean up to remove all unnecessary special case checking in the constructors. Right now this example should work (which also becomes part of the `test_factory_zero_nnz` test).